File: data_structure/priorityQueue.c
#include <stdio.h>
#include <stdlib.h>

char* error = "None";

struct Node {
  int data;
  struct Node* link;
};

struct Node* head;

void pop() {
  if (head == NULL) {
    error = "Linked list is empty";
    return;
  }

  head = head->link;
}

void printError() {
  printf("Error: %s", error);
}

void push(int _data) {
  struct Node* node = (struct Node*)malloc(sizeof(struct Node));
  node->data = _data;

  struct Node *current = head, *previous = NULL;
  while (current != NULL) {
    if (current->data > _data)
      break;

    previous = current;
    current = current->link;
  }

  if (previous == NULL) {
    node->link = head;
    head = node;
  } else {
    node->link = previous->link;
    previous->link = node;
  }
}

void traverse() {
  if (head == NULL) {
    error = "Linked list is empty";
    return;
  }

  printf("\nLinked list: ");
  struct Node* current = head;
  while (current != NULL) {
    printf("%d ", current->data);
    current = current->link;
  }
}

int main() {
  while (1) {
    system("clear");
    int choice, data;

    printError();
    traverse();

    printf("\nMenu:\n1) Add data\n2) Pop data\n3) Exit\nEnter choice(1/2/3): ");
    scanf("%d", &choice);

    switch (choice) {
      case 1:
        printf("Enter data: ");
        scanf("%d", &data);
        push(data);
        break;

      case 2:
        pop();
        break;

      case 3:
        printf("Exited!\n");
        exit(0);

      default:
        error = "Invalid choice";
    }
  }

  return 0;
}

File: data_structure/PriorityQueue.java
package data_structure;

import java.util.Scanner;

class Node {
  int data;
  int priority;
  Node link;

  Node(int data, int priority) {
    this.data = data;
    this.priority = priority;
    this.link = null;
  }
}

public class PriorityQueue {
  static Node head = null;
  static Node tail = null;
  static String error = "None"; 

  static void clrscr() {
    System.out.print("\033[H\033[2J");
    System.out.flush();
  }

  static void error() {
    System.out.println("Error: " + error);
  }

  static void pop() {
    if (head == null) {
      error = "Linked list is empty";
      return;
    }

    if (head.link == null)
      tail = null;

    head = head.link;
    error = "None";
  }

  static void push(int data, int priority) {
    Node node = new Node(data, priority);
    Node current = head;
    boolean before = true;
 
    if (current != null) {
      if (current == tail) {
        if (current.data < priority)
          current = null;
      } else {
        boolean atFirst = true;
        while (current.link != null) {
          if (atFirst) {
            if (current.data > priority)
              break;

            atFirst = false;
            continue;
          } 

          if (current.link.data > priority) {
            before = false;
            break;
          }

          current = current.link;
        }
      }
    }

    if (head == null)
      head = tail = node;
    else if (current == head && before) {
      node.link = current;
      head = node;
    } else if (current != null && current.link != null) {
      node.link = current.link;
      current.link = node;
    } else {
      tail.link = node;
      tail = node;
    }

    error = "None";
  }

  static void traverse() {
    System.out.print("Linked list: ");
    if (head == null) {
      error = "Linked list is empty";
      return;
    }
    
    Node current = head;
    while (current != null) {
      System.out.print(current.data + " ");
      current = current.link;
    }
  }

  public static void main(String[] args) {
    Scanner scanner = new Scanner(System.in);
    boolean flag = true;
    while (flag) {
      clrscr();
      error();
      traverse();

      System.out.print("\nMenu:\na) Push data\nb) Pop data\nc) Exit\nEnter choice (a/b/c): ");
      int choice = scanner.next().charAt(0);
      int data;
      switch (choice) {
        case 'a':
          System.out.print("Enter data: ");
          data = scanner.nextInt();
          System.out.print("Enter priority: ");
          push(data, scanner.nextInt());
          break;

        case 'b':
          pop();
          break;

        case 'c':
          System.out.print("Exited!");
          flag = false;
          break;

        default:
          error = "Invalid choice";
      }
    }
    scanner.close();
  }
}
